Improve BC42328 documentation with comprehensive explanation of relaxed delegate conversions (#48120)
dotnet/docs: docs/visual-basic/misc/bc42328.md

ms.assetid: 1561e5bf-b8ab-4a67-990d-c3a76e2ab353
---
# The 'AddressOf' expression has no effect in this context because the method argument to 'AddressOf' requires a relaxed conversion to the delegate type of the event

> This warning occurs when you use `AddressOf` with a method in an event handler context where the method signature doesn't exactly match the event's delegate signature. Visual Basic allows "relaxed delegate conversion" in these scenarios, but the `AddressOf` operator becomes redundant because the compiler must create a wrapper delegate anyway.

Reviewer: ```suggestion
This warning occurs when you use `AddressOf` with a method in an event handler context where the method signature doesn't exactly match the event's delegate signature. Visual Basic allows *relaxed delegate conversion* in these scenarios, but the `AddressOf` operator becomes redundant because the compiler must create a wrapper delegate anyway.
```

> ## Understanding relaxed delegate conversion

Reviewer 2: avoid gerund form:

```suggestion
## Understand relaxed delegate conversion
```

> Visual Basic allows you to assign methods to delegates even when their signatures don't match exactly, as long as the conversion is safe. This is called "relaxed delegate conversion" and includes:

Reviewer: ```suggestion
Visual Basic allows you to assign methods to delegates even when their signatures don't match exactly, as long as the conversion is safe. This is called *relaxed delegate conversion* and includes:
```

> - **Widening parameter types**: A method parameter can be a wider type than the delegate parameter
> - **Widening return types**: A method return type can be narrower than the delegate return type  
> - **Omitting parameters**: A method can have fewer parameters than the delegate
> - **Dropping return values**: A function can be assigned to a `Sub` delegate

Reviewer 2: Add periods on each bullet:

```suggestion
- **Widening parameter types**: A method parameter can be a wider type than the delegate parameter.
- **Widening return types**: A method return type can be narrower than the delegate return type.  
- **Omitting parameters**: A method can have fewer parameters than the delegate.
- **Dropping return values**: A function can be assigned to a `Sub` delegate.
```

> - **Widening parameter types**: A method parameter can be a wider type than the delegate parameter.
> - **Widening return types**: A method return type can be narrower than the delegate return type.  
> - **Omitting parameters**: A method can have fewer parameters than the delegate.
> - **Dropping return values**: A function can be assigned to a `Sub` delegate.

Reviewer: ```suggestion
- **Widened parameter types**: A method parameter can be a wider type than the delegate parameter.
- **Widened return types**: A method return type can be narrower than the delegate return type.  
- **Omitted parameters**: A method can have fewer parameters than the delegate.
- **Dropped return values**: A function can be assigned to a `Sub` delegate.
```

Reviewer: Is this better, so to avoid -ing words?

## When this warning occurs

This warning appears when Visual Basic can handle the assignment through relaxed conversion, making the explicit `AddressOf` unnecessary. For example:

```vb
' Event delegate expects (sender As Object, e As EventArgs)
Public Event SomeEvent As EventHandler

' Method with exact signature - no warning
Sub Handler1(sender As Object, e As EventArgs)
    ' Implementation
End Sub

' Method with omitted parameters - causes BC42328
Sub Handler2()
    ' Implementation  
End Sub

' Using AddHandler with exact match - no warning
AddHandler SomeEvent, AddressOf Handler1

' Using AddHandler with relaxed conversion - causes BC42328
AddHandler SomeEvent, AddressOf Handler2
```

## Why AddressOf appears to have "no effect"

> The warning message states that `AddressOf` "has no effect" because:
> 
> 1. Visual Basic must create a wrapper delegate for the relaxed conversion regardless
> 2. The `AddressOf` operator doesn't change how the compiler handles the conversion
> 3. The same result occurs whether you use `AddressOf` or not in relaxed conversion scenarios

Reviewer: I couldn't tell if 1-3 were mutually exclusive, or *all* true. Maybe this is clearer?

```suggestion
The warning message states that `AddressOf` "has no effect" because Visual Basic must create a wrapper delegate for the relaxed conversion regardless. The `AddressOf` operator doesn't change how the compiler handles the conversion: the same result occurs whether you use `AddressOf` or not in relaxed conversion scenarios.
```

However, `AddressOf` is still syntactically required in `AddHandler` and `RemoveHandler` statements.

**Error ID:** BC42328

## To correct this error

You have several options depending on your needs:

> 1. **Assign to a variable first (preserves exact semantics):**
> 
>    ```vb
>    Dim handler As EventHandler = AddressOf Handler2
>    AddHandler SomeEvent, handler
>    ```
> 
> 1. **Match the delegate signature exactly:**
> 
>    ```vb
>    Sub Handler2(sender As Object, e As EventArgs)
>        ' Implementation - use parameters or ignore them
>    End Sub
>    ```
> 
> 1. **Use a lambda expression:**
> 
>    ```vb
>    AddHandler SomeEvent, Sub() Handler2()
>    ```
> 
> 1. **Use the `Handles` clause if appropriate:**
> 
>    ```vb
>    Sub Handler2() Handles someObject.SomeEvent
>        ' Implementation
>    End Sub
>    ```

Reviewer 2: These should not be numbered, as they are "one of" options, not sequential steps:

````suggestion
- **Assign to a variable first (preserves exact semantics):**

   ```vb
   Dim handler As EventHandler = AddressOf Handler2
   AddHandler SomeEvent, handler
   ```

- **Match the delegate signature exactly:**

   ```vb
   Sub Handler2(sender As Object, e As EventArgs)
       ' Implementation - use parameters or ignore them
   End Sub
  ```

- **Use a lambda expression:**

   ```vb
   AddHandler SomeEvent, Sub() Handler2()
   ```

- **Use the `Handles` clause if appropriate:**

   ```vb
   Sub Handler2() Handles someObject.SomeEvent
       ' Implementation
   End Sub
   ```
````

## When you can ignore this warning

This warning is often safe to ignore when:

> - Your handler method intentionally omits parameters it doesn't need
> - You're using a simpler method signature for cleaner code
> - The performance impact is negligible for your application

Reviewer: ```suggestion
- Your handler method intentionally omits parameters it doesn't need.
- You're using a simpler method signature for cleaner code.
- The performance impact is negligible for your application.
```
